From saharabeara@gmail.com  Fri Feb 23 10:16:00 2007
From: Sarah Bailey <saharabeara@gmail.com>
Date: Thu, 22 Feb 2007 22:36:21 -0800
Subject: gadgetfs: Fixed bug in ep_aio_read_retry.
To: David Brownell <david-b@pacbell.net>, greg@kroah.com
Cc: linux-usb-devel@lists.sourceforge.net, usb-hacking@svcs.cs.pdx.edu
Message-ID: <20070223063621.GA20539@localdomain>
Content-Disposition: inline


I don't think the current code works with multiple iovecs.
The original would just copy the first part of priv->buf
over and over into multiple iovecs.

Signed-off-by: Sarah Bailey <saharabeara@gmail.com>
Signed-off-by: David Brownell <dbrownell@users.sourceforge.net>

---
 drivers/usb/gadget/inode.c |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

--- gregkh-2.6.orig/drivers/usb/gadget/inode.c
+++ gregkh-2.6/drivers/usb/gadget/inode.c
@@ -553,6 +553,7 @@ static ssize_t ep_aio_read_retry(struct 
 {
 	struct kiocb_priv	*priv = iocb->private;
 	ssize_t			len, total;
+	void			*to_copy;
 	int			i;
 
 	/* we "retry" to get the right mm context for this: */
@@ -560,10 +561,11 @@ static ssize_t ep_aio_read_retry(struct 
 	/* copy stuff into user buffers */
 	total = priv->actual;
 	len = 0;
+	to_copy = priv->buf;
 	for (i=0; i < priv->nr_segs; i++) {
 		ssize_t this = min((ssize_t)(priv->iv[i].iov_len), total);
 
-		if (copy_to_user(priv->iv[i].iov_base, priv->buf, this)) {
+		if (copy_to_user(priv->iv[i].iov_base, to_copy, this)) {
 			if (len == 0)
 				len = -EFAULT;
 			break;
@@ -571,6 +573,7 @@ static ssize_t ep_aio_read_retry(struct 
 
 		total -= this;
 		len += this;
+		to_copy += this;
 		if (total == 0)
 			break;
 	}
